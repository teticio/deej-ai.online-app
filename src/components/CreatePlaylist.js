import React, { useState } from 'react';
import { Text, Card, Spinner, Hr, FaForward, FaCog } from './Platform';
import { VerticalSpacer, Row } from './Lib';
import AddTrack from './AddTrack';
import RemovablePlaylist from './RemovablePlaylist';
import generatePlaylist from './GeneratePlaylist';
import SavePlaylist from './SavePlaylist';

export default function CreatePlaylist({
  waypoints = { track_ids: [] },
  size = 10,
  creativity = 0.5,
  noise = 0,
  onCreate = f => f,
  onSettings = f => f,
  spotify = null,
  savePlaylist = true
}) {
  const [_waypoints, setWaypoints] = useState(waypoints);
  const [spinner, setSpinner] = useState(false);

  const handleClick = () => {
    setSpinner(true);
    generatePlaylist(_waypoints, size, creativity, noise)
      .then(playlist => {
        if (savePlaylist) {
          SavePlaylist(playlist, creativity, noise)
            .then(id => playlist.id = id);
        }
        return playlist;
      })
      .then(playlist => onCreate(playlist, _waypoints))
      .catch(error => console.error('Error:', error));
  }

  return (
    <>
      <Text h4 style={{ textAlign: 'center' }}>Choose the waypoints for your musical journey</Text>
      <AddTrack
        numTracks={_waypoints.track_ids.length}
        spotify={spotify}
        onAdd={id => {
          setWaypoints({ 'track_ids': _waypoints.track_ids.concat(id) });
        }} />
      <VerticalSpacer />
      <Card style={{ padding: 15 }}>
        <RemovablePlaylist {..._waypoints} onRemove={id => {
          setWaypoints({ 'track_ids': _waypoints.track_ids.filter((element, index) => index !== id) });
        }} />
        {_waypoints.track_ids.length > 0 ?
          <Hr /> :
          <></>
        }
        {spinner ?
          <Spinner animation='border' size='md' /> :
          <Row style={{ justifyContent: 'space-between'}} surface={true} >
            <FaForward
              data-testid='create-playlist'
              size='25'
              className='link'
              onClick={handleClick}
            />
            <FaCog
              data-testid='settings'
              size='25'
              className='link'
              onClick={() => onSettings(_waypoints)}
            />
          </Row>
        }
      </Card>
    </>
  );
}
