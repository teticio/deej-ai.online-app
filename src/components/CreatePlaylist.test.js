import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import CreatePlaylist from './CreatePlaylist';

jest.setTimeout(20000);

test('playlist creation', async () => {
  const onCreate = jest.fn();

  render(
    <CreatePlaylist
      waypoints={{
        track_ids: [
          '1O0xeZrBDbq7HPREdmYUYK',
          '1b7LMtXCXGc2EwOIplI35z'
        ]
      }}
      onCreate={onCreate}
      savePlaylist={false}
    />
  );
  fireEvent.click(screen.getByTestId('create-playlist'));
  await waitFor(() => expect(onCreate).toHaveBeenCalledTimes(1), { timeout: 20000 });
  expect(onCreate).toMatchSnapshot();
});
