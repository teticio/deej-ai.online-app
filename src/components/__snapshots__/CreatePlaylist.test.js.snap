// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`playlist creation 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "av_rating": 0,
        "name": "Deej-A.I.",
        "num_ratings": 0,
        "track_ids": Array [
          "1O0xeZrBDbq7HPREdmYUYK",
          "6Y0ed41KYLRnJJyYGGaDgY",
          "5yrsBzgHkfu2idkl2ILQis",
          "6yXcmVKGjFofPWvW9ustQX",
          "1DKyFVzIh1oa1fFnEmTkIl",
          "6b8hjwuGl1H9o5ZbrHJcpJ",
          "5qRJD1yaLJ5s0J3JpbgnwA",
          "6kotXaSQaGYxE62hVpdHWu",
          "4lrQv8z3qq1Rl8bsc0Qy0y",
          "2nmaEzFZrSm2aMLtfJDzyG",
          "3PPDUkGHUJx2bxct6A3PBy",
          "1b7LMtXCXGc2EwOIplI35z",
        ],
        "tracks": Array [
          "Jim Croce - Bad, Bad Leroy Brown",
          "The Animals - We've Gotta Get Out of This Place",
          "Jim Croce - You Don't Mess Around with Jim",
          "Redbone - The Witch Queen of New Orleans",
          "Them - Gloria - Stereo Version",
          "Yusuf / Cat Stevens - Hard Headed Woman",
          "Electric Light Orchestra - Evil Woman",
          "Warren Zevon - Werewolves Of London - 2007 Remastered Version",
          "Electric Light Orchestra - Mr. Blue Sky",
          "Warren Zevon - Werewolves Of London",
          "Norman Greenbaum - Spirit In The Sky",
          "Stealers Wheel - Stuck In The Middle With You",
        ],
        "waypoints": Array [
          "1O0xeZrBDbq7HPREdmYUYK",
          "1b7LMtXCXGc2EwOIplI35z",
        ],
      },
      Object {
        "track_ids": Array [
          "1O0xeZrBDbq7HPREdmYUYK",
          "1b7LMtXCXGc2EwOIplI35z",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
